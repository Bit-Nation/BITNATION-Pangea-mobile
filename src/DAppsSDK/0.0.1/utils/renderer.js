// @flow

import React from 'react';
import components from '../components';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => components[typeName];

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, custom: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const customProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (typeof propName !== 'string') {
      console.warn('Prop name is not a string');
      return;
    }

    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (validProps.custom.includes(propName)) {
      // It's an allowed custom prop.
      customProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, customProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === components.Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = components[type];

  if (component == null) {
    console.warn(`Trying to render unknown component type ${type}.`);
    return null;
  }

  if (component.validNativeProps != null && Array.isArray(component.validNativeProps) === false) {
    console.warn(`validNativeProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.customProps != null && Array.isArray(component.customProps) === false) {
    console.warn(`customProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.callbackProps != null && Array.isArray(component.callbackProps) === false) {
    console.warn(`callbackProps field of component ${type} should be an array of strings.`);
    return null;
  }

  const { props } = json;

  const { nativeProps = props, customProps = {}, callbackProps = {} } =
    validateProps(json.props, {
      native: component.validNativeProps || [],
      custom: component.customProps || [],
      callbacks: component.callbackProps || [],
    }, type);

  return React.createElement(
    component,
    {
      nativeProps,
      ...customProps,
      ...customPropsProvider(component, { ...nativeProps, ...customProps, ...callbackProps }),
      key,
    },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
