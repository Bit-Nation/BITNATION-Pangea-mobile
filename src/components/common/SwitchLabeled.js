// @flow

import React from 'react';
import { Switch, View, Text } from 'react-native';
import { MediaQueryStyleSheet } from 'react-native-responsive';
import GlobalStyles from '../../global/Styles';
import Colors from '../../global/colors';

type Props = {
  /**
   * @desc Value of the Switch.
   */
  value?: boolean,
  /**
   * @desc Text label at the right of the Switch.
   */
  label?: string,
  /**
   * @desc Switch alignment
   */
  align?: string,
  /**
   * @desc Callback to be called on Switch when changes value.
   */
  onValueChange?: () => void,
  /**
   * @desc Swich disabled status
   */
  disabled?: boolean,
};

/**
 * @desc Generates a standard native Switch with a Label at the right
 * @return {React.Component} A component.
 */

const SwitchLabeled = ({
  onValueChange, value, label, align, disabled,
}:Props) => {
  const styles = MediaQueryStyleSheet.create({
    ...GlobalStyles,
  });

  const switchStyle = [styles.switchContainer];
  if (align === 'right') {
    switchStyle.push({
      flexDirection: 'row-reverse',
    });
  }

  const textStyle = [styles.formSwitchLabelText];
  if (disabled) {
    textStyle.push(styles.disabledText);
  }

  return (
    <View style={styles.formRow}>
      <View style={switchStyle}>
        <Switch
          testID='account_developerSetting_switchTestingAccount'
          disabled={disabled}
          style={styles.switchObject}
          onTintColor={Colors.BitnationHighlightColor}
          onValueChange={onValueChange}
          value={value}
        />
        <Text style={textStyle}>{label}</Text>
      </View>
    </View>
  );
};

SwitchLabeled.defaultProps = {
  value: false,
  label: '',
  align: 'left',
  disabled: false,
  onValueChange: () => undefined,
};

export default SwitchLabeled;
