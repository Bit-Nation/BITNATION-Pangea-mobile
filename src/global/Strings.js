/*
	Bitnation Text Strings
	Version 0.3.1
	
	Text strings for use in the UI, and to represent coded values.
	
	Notes:
	- Some strings have both a regular and Title version, one with Title Case the other with Sentence case.
*/

// ========================================
// DEFAULT STRINGS
// ========================================

const defaultStrings = {
	
	//General for UI
	optional: 'Optional',
	options: 'Options',
	fees: 'Fees',
	ok: 'OK',
	
	nation: 'nation',
	nations: 'nations',
	
	// Navigation
	dashboard: 'Dashboard',
	nations: 'Nations',
	chat: 'Chat',
	wallet: 'Wallet',
	profile: 'Profile',
	
	// Nations
	createNationTitle: 'Create A Nation',
	createNationIntroText: 'Reclaim your sovereignty by creating your own decentralized borderless voluntary nation(DBVN).',
	createNationNationNamePrompt: 'Nation name',
	createNationShortDescPrompt: 'Short description',
	createNationLocationPrompt: 'Nation location',
	createNationRepresentsPrompt: 'Nation represents an existing Earth country',
	createNationCoreFootnote: 'Virtual nations exist only in Pangea. Geographic nations are on Earth.',
	createNationNoLegalCode: 'No legal code',
	createNationLegalCode: 'Legal Code...',
	createNationGovernmentalStructure: 'Governmental Structure',
	
	createNationLawEnforcementMechanism: 'Law Enforcement Mechanism...',
	createNationNoLawEnforcement: 'No Law Enforcement',
	createNationIDReputationLawEnforcement: 'ID & Reputation Law Enforcement',
	createNationPrivateOrCooperativeSecurity: 'Private or Cooperative Security',
	createNationNationStateLawEnforcement: 'Nation State Law Enforcement',
	createNationInternationalLawEnforcement: 'International Law Enforcement',
	
	createNationTypeOfGovernment: 'Type of Government...',
	createNationOtherTypeOfGovernment: 'Other Type of Government',
	createNationHolocracy: 'Holocracy',
	createNationDemocracy: 'Democracy',
	createNationAutocracy: 'Autocracy',
	createNationTheocracy: 'Theocracy',
	createNationMeritocracy: 'Meritocracy',
	createNationsServicesOffered: 'Services Offered...',
	createNationsLegalServices: 'Legal Services',
	createNationsInsuranceServices: 'Insurance Services',
	createNationsSocialServices: 'Social Services',
	createNationsDiplomaticServices: 'Diplomatic Services',
	createNationsPhysicalServices: 'Physical Services',
	createNationsSeekingDiplomaticRecognition: 'Seeking diplomatic recognition as a sovereign entity.',
	createNationsNonCitizensMayUseGovernanceServices: 'Non-citizens may use governance services.',
	createNationForProfitNation: 'For-profit nation.',
	
	// Nation View Detail
	nationDetailLocationMsg: '[[nationName]] is a [[locationType]] nation.',
	
	governmentalStructureFootnote: '',
	
	// Nations generic strings
	governmentalStructure: 'Governmental structure',
	governmentalStructureTitle: 'Governmental Structure',
	legalCode: 'Legal code',
	
	// Legal Codes
	legalCodeTransSupra: 'Transnational/Supranational Law',
	legalCodeMultiple: 'Multiple Legal Codes',
	legalCodeUNIDROIT: 'UNIDROIT Principles',
	legalCodeComputer: 'Computer Code',
	legalCodeCommon: 'Common Law',
	legalCodeCivil: 'Civil Law',
	
	// Nations Fields
	virtualNationTrue: 'Virtual',
	virtualNationFalse: 'Geographical',
	
	// Wallet strings
	send: 'Send',
	receive: 'Receive',
	sendTo: 'To',
	receiveFrom: 'From',
	note: 'Note',
	amount: 'Amount',
	fee: 'Fee',
	blockchainAddress: 'Address',
	
	sendCommand: 'Send',
	
}

// ========================================
// Nation Strings
// ========================================
const walletStrings = {
	
	scanQRCodeMsg: 'The sender can scan this QR code with a phone or computer camera to get your wallet address.',
	copyAddressMsg: 'You can copy your wallet address and send any way you choose, e.g. SMS or email. Do not try to type your address by hand!',
	
  walletUpdating: 'Updating',
  walletUpdateFailed: 'Update failed',
}

// ========================================
// Wallet Strings
// ========================================
const nationStrings = {
	scanQRCodeMsg: 'The sender can scan this QR code with a phone or computer camera to get your wallet address.',
	copyAddressMsg: 'You can copy your wallet address and send any way you choose, e.g. SMS or email. Do not try to type your address by hand!',
}

// ========================================
// Dashboard Strings
// ========================================
const dashboardStrings = {
  warningBody: 'Pangea is in an early development stage, use with caution and at your own risk.',
	warningTitle: 'Warning',
	walletEmptyState: 'Begin using Pangea by creating your wallet.',
}

// ========================================
// LANGUAGE LOCALIZATION STRINGS
// ========================================

const FR = {
	governmentalStructure: 'Structure gouvernementale',
}

// Round currency for screen display. Different currencies might require different treatment.
// currency (string) The code of the currency, e.g. 'ETH' or 'BTC'
// amount (real) The amount of currency to show, e.g. 0.21495893984
// addCurrency (bool) True = format the display string with the currency name (might be different from 'currency', above!)
//                      false = return rounded currency without the name
const showCurrency = (currency, amount, addCurrency) => {
	if (!amount) {
		amount = 0
	}
	// pow is the
	var precision, pattern
	
	switch (currency) {
		case 'ETH':
			precision = 5
			pattern = '{amount} ETH'
			break
		default:
			precision = 5
	}
	// Current default is ETH, round to 1/10,000
	var factor = Math.pow(10, precision)
	amount = Math.round(amount * factor) / factor
	
	var displayAmount = amount.toFixed(precision)
	if (addCurrency) {
		displayAmount = pattern.replace('{amount}', displayAmount)
	}
	return displayAmount
}

const stringFunctions = {
	showCurrency: showCurrency,
}

const strings = {
	...defaultStrings,
	...walletStrings,
	...nationStrings,
	...stringFunctions,
	dashboard: dashboardStrings,
}

export default strings
