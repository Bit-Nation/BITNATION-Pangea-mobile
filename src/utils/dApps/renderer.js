// @flow

import React from 'react';

import Text from '../../components/dApps/Text';
import View from '../../components/dApps/View';
import TextInput from '../../components/dApps/TextInput';
import Button from '../../components/dApps/Button';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => {
  switch (typeName) {
    case 'Text':
      return Text;
    case 'View':
      return View;
    case 'TextInput':
      return TextInput;
    case 'Button':
      return Button;
    default:
      console.warn(`Unknown component type ${typeName}.`);
      return null;
  }
};

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, stateBased: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const stateBasedProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (propName.endsWith('Path') && validProps.stateBased.includes(propName.substring(0, propName.length - 4))) {
      // It's an allowed prop to set or get root component state. That properties are marked by Path suffix.
      stateBasedProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, stateBasedProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = getTypeElementFromText(type);

  if (component === null) {
    return null;
  }

  const { props } = json;

  const { nativeProps = props, stateBasedProps = {}, callbackProps = {} } =
    component.validNativeProps !== undefined
      ? validateProps(json.props, {
        native: component.validNativeProps,
        stateBased: Object.keys(component.stateBasedProps || {}),
        callbacks: component.callbackProps || [],
      }, type)
      : {};

  return React.createElement(
    component,
    { ...nativeProps, ...customPropsProvider(component, { ...nativeProps, ...stateBasedProps, ...callbackProps }), key },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
