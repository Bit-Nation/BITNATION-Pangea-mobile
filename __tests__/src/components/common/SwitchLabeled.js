import SwitchLabeled from '../../../../src/components/common/SwitchLabeled';
import React from 'react';
import renderer from 'react-test-renderer'
import Strings from "../../../../src/global/Strings";

test('SwitchLabeled renders correctly', () => {
  const tree = renderer.create(
    <SwitchLabeled
      label={'Testing SwitchLabeled'}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('SwitchLabeled renders correctly with set value', () => {
  const tree = renderer.create(
    <SwitchLabeled
      label={'Testing SwitchLabeled'}
      value={true}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});