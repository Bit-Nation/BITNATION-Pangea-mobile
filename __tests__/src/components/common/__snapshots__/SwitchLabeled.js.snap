// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchLabeled component tests SwitchLabeled rendering SwitchLabeled renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderWidth": 0,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
          "marginTop": 5,
        },
      ]
    }
  >
    <RCTSwitch
      disabled={false}
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onTintColor="#F5A623"
      onValueChange={[Function]}
      style={
        Array [
          Object {
            "height": 31,
            "width": 51,
          },
          Object {
            "marginRight": 10,
          },
        ]
      }
      testID="account_developerSetting_switchTestingAccount"
      value={false}
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#58595B",
            "fontFamily": "Roboto",
            "fontSize": 16,
            "fontWeight": "normal",
            "letterSpacing": 0,
            "lineHeight": 22,
            "marginRight": 60,
          },
        ]
      }
    >
      Testing SwitchLabeled
    </Text>
  </View>
</View>
`;

exports[`SwitchLabeled component tests SwitchLabeled rendering SwitchLabeled renders correctly with set value 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderWidth": 0,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
          "marginTop": 5,
        },
      ]
    }
  >
    <RCTSwitch
      disabled={false}
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onTintColor="#F5A623"
      onValueChange={[Function]}
      style={
        Array [
          Object {
            "height": 31,
            "width": 51,
          },
          Object {
            "marginRight": 10,
          },
        ]
      }
      testID="account_developerSetting_switchTestingAccount"
      value={true}
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#58595B",
            "fontFamily": "Roboto",
            "fontSize": 16,
            "fontWeight": "normal",
            "letterSpacing": 0,
            "lineHeight": 22,
            "marginRight": 60,
          },
        ]
      }
    >
      Testing SwitchLabeled
    </Text>
  </View>
</View>
`;
